ERC721 Docs (#117)
OpenZeppelin/cairo-contracts: docs/ERC721.md

- `ERC165_register_interface` allows contracts to set and communicate which interfaces they support. This follows OpenZeppelin's [ERC165Storage](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/introspection/ERC165Storage.sol)
- `IERC721_Receiver` compliant contracts (`ERC721_Holder`) return a hardcoded selector id according to EVM selectors, since selectors are calculated differently in Cairo. This is in line with the ERC165 interfaces design choice towards EVM compatibility
- `IERC721_Receiver` includes the additional method `is_account` in order to determine if the recipient contract is an account contract. The Solidity implementation differs in that externally owned accounts (EOAs) could easily be distinguished from smart contracts. StarkNet currently does not have a differentiator between accounts and non-accounts; therefore, contracts that support safe ERC721 transfers must return `0` from `is_account` in order to communicate that the recipient contract is not an account but does support ERC721 transfers

> - ERC721_Enumerable tracks the total number of tokens by using the `all_tokens` mapping and `all_tokens_len` storage variable which together mimic an array in Cairo. The Solidity implementation instead uses an array

Reviewer: ```suggestion
- `ERC721_Enumerable` tracks the total number of tokens with the `all_tokens` and `all_tokens_len` storage variables mimicking the array of the Solidity implementation.
```

## Usage

Use cases go from artwork, digital collectibles, physical property, and many more.

> To show a standard use-case, we'll use the `ERC721_Mintable` preset which allows for only the owner to `mint` and `burn` tokens. To create a token you need to first deploy both Account and ERC721 contracts respectively. As most StarkNet contracts, ERC721 expects to be called by another contract and it identifies it through `get_caller_address` (analogous to Solidity's `this.address`). This is why we need an Account contract to interact with it.

Reviewer: ```suggestion
To show a standard use case, we'll use the `ERC721_Mintable` preset which allows for only the owner to `mint` and `burn` tokens. To create a token you need to first deploy both Account and ERC721 contracts respectively. As most StarkNet contracts, ERC721 expects to be called by another contract and it identifies it through `get_caller_address` (analogous to Solidity's `this.address`). This is why we need an Account contract to interact with it.
```

StarkNet contracts that support safe transfers, however, must also include `is_account`. `safeTransferFrom` requires a means of differentiating between account and non-account contracts. Currently, StarkNet does not support error handling from the contract level; therefore, the current ERC721 implementation requires that all contracts that support safe ERC721 transfers (both accounts and non-accounts) include the `is_account` method. This method should return `0` since it's NOT an account. 

As Cairo/StarkNet matures and more features are added, new strategies will be examined to perhaps omit the `is_account` method from contracts that support safe ERC721 transfers.

#### IERC721_Receiver

Interface for any contract that wants to support safeTransfers from ERC721 asset contracts.

```jsx
@contract_interface
namespace IERC721_Receiver:
    func onERC721Received(
        operator: felt,
        _from: felt,
        tokenId: Uint256,
        data_len: felt
        data: felt*
    ) -> (selector: felt):
    end

    func is_account() -> (res: felt):
    end
end
```

### Supporting Interfaces 

> In order to ensure EVM/StarkNet compatibility, this ERC721 implementation does not calculate interface identifiers. Instead, the interface IDs are hardcoded from their Ethereum Virtual Machine (EVM) calculations. With the EVM, the interface ID is calculated from the selector's first four bytes of the hash of the function's signature; however, Cairo's protocol for calculating selectors includes 252 bytes. Due to this difference, hardcoding EVM's already-calculated interface IDs is the most consistent approach to both follow the EIP165 standard and EVM compatibility.

Reviewer: ```suggestion
In order to ensure EVM/StarkNet compatibility, this ERC721 implementation does not calculate interface identifiers. Instead, the interface IDs are hardcoded from their EVM calculations. On the EVM, the interface ID is calculated from the selector's first four bytes of the hash of the function's signature; however, Cairo's protocol for calculating selectors includes 252 bytes. Due to this difference, hardcoding EVM's already-calculated interface IDs is the most consistent approach to both follow the EIP165 standard and EVM compatibility.
```

#### IERC721_Metadata

```jsx
@contract_interface
namespace IERC721_Metadata:
    func name() -> (name: felt):
    end

    func symbol() -> (symbol: felt):
    end

    func tokenURI(tokenId: Uint256) -> (tokenURI: felt):
    end
end
```

### ERC721_Enumerable

The Enumerable extension adds enumerability of all the token ids in the contract as well as all token ids owned by each account. This allows contracts to publish its full list of NFTs and make them discoverable. 

In regard to implementation, frontend contracts should import the following view methods:
-  `ERC721_Enumerable_totalSupply`
-  `ERC721_Enumerable_tokenByIndex`
-  `ERC721_Enumerable_tokenOfOwnerByIndex`

In order for the tokens to be correctly indexed, the frontend contract should also import the following methods (which supercede some of the ERC721_base methods):
-  `ERC721_Enumerable_transferFrom`
-  `ERC721_Enumerable_safeTransferFrom`
-  `ERC721_Enumerable_mint`
-  `ERC721_Enumerable_burn`

> ### IERC721_Enumerable

Reviewer: ```suggestion
#### IERC721_Enumerable
```

  * [`IERC165`](#ierc165)
    - [`supportsInterface`](#-supportsInterface-)

## Interfaces

### IERC721

```jsx
@contract_interface
namespace IERC721:
    func balanceOf(owner: felt) -> (balance: Uint256):
    end

    func ownerOf(token_id: Uint256) -> (owner: felt):
    end

    func safeTransferFrom(
        _from: felt, 
        to: felt, 
        token_id: Uint256, 
        data_len: felt,
        data: felt*
    ):

    func transferFrom(_from: felt, to: felt, token_id: Uint256):
    end

    func approve(approved: felt, token_id: Uint256):
    end

    func setApprovalForAll(operator: felt, approved: felt):
    end

    func getApproved(token_id: Uint256) -> (approved: felt):
    end

    func isApprovedForAll(owner: felt, operator: felt) -> (is_approved: felt):
    end

--------------- IERC165 ---------------

    func supportsInterface(interface_id: felt) -> (success: felt):
    end

```

### IERC721_Metadata

```jsx
@contract_interface
namespace IERC721_Metadata:
    func name() -> (name: felt):
    end

    func symbol() -> (symbol: felt):
    end

    func tokenURI(token_id: Uint256) -> (uri_len: felt, uri: felt*):
    end

> --------------- IERC721 ---------------

Reviewer: maybe we can indent this 😬

    end

    func approve(approved: felt, token_id: Uint256):
    end

    func setApprovalForAll(operator: felt, approved: felt):
    end

    func getApproved(token_id: Uint256) -> (approved: felt):
    end

    func isApprovedForAll(owner: felt, operator: felt) -> (is_approved: felt):
    end

--------------- IERC165 ---------------

    func supportsInterface(interface_id: felt) -> (success: felt):
    end

```

### IERC721_Receiver

```jsx
@contract_interface
namespace IERC721_Receiver:
    func onERC721Received(
        operator: felt,
        _from: felt,
        token_id: Uint256,
        data_len: felt
        data: felt*
    ) -> (selector: felt):
    end
```

### IERC165
```jsx
@contract_interface
namespace IERC165:
    func supportsInterface(interface_id: felt) -> (success: felt):
    end
```

### ERC721 Compatibility

Although StarkNet is not EVM compatible, this implementation aims to be as close as possible to the ERC721 standard in the following ways:

- it uses Cairo's `uint256` instead of `felt`
- it returns `1` as true/success to imitate a `bool`
- it makes use of Cairo's short strings to simulate `name` and `symbol`

> - it will emit events once they're implemented on StarkNet

Reviewer: Some of this we might want to abstract away from here and ERC20 docs, then link to it since it will probably change over time.

1. if the calling address is an externally owned account (EOA), the token transfer will behave as if `transferFrom` was called
2. if the calling address is a contract, the safe function will check that the contract supports ERC721 tokens

The current implementation of `safeTansferFrom` checks for `onERC721Received`. 
Because EOAs cannot be differentiated from regular contracts on StarkNet at this time, it's recommended to transfer tokens with `transferFrom`; however, please be aware that transfering tokens to a contract that does not support ERC721 can result in lost tokens forever. 

### ERC721Received

In order to be sure a contract can safely accept ERC721 tokens, said contract must implement the `ERC721_Receiver` interface. Methods such as `safeTransferFrom` and `safeMint` call the recipient contract's `onERC721Received` method. If the contract fails to return the correct magic value, the transaction fails. 

> Please note: EOA/contract differentiation must be resolved in order to correctly use this feature

### Supporting Interfaces

> In order to attain seamless layer one interoperability, this ERC721 implementation uses hardcoded interface IDs that have already been calculated on layer one. The Solidity implementation of ERC165 uses the `super` method to amend parent functions. Being that Cairo/Starknet currently does not support this feature, this implementation uses a mapping to store supported interfaces (similar to OpenZeppelin's [ERC165Storage](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/introspection/ERC165Storage.sol)). 

Reviewer: I think the focus is around Cairo <> EVM compatibility rather than Cairo <> Solidity implementation details.

## Extensibility

> This implementation of Cairo smart contract extensibility is set up to include all ERC721 storage and function logic within the base contract `ERC721_base.cairo`. Base contract methods with the prefix `ERC721_` must be imported to the frontend contract and inserted into an `external` method with the requisite name. This is already done in `ERC721.cairo`; however, additional functionality can be added. For instance, you could:

Reviewer: I think this reads better and it also links to the pattern, otherwise people won't know what extensibility are we talking about.
```suggestion
Following the [contracts extensibility pattern](https://community.starknet.io/t/contract-extensibility-pattern/210), this implementation is set up to include all ERC721 storage and function logic within the base contract `ERC721_base.cairo`. Base contract methods with the prefix `ERC721_` must be imported to the frontend contract and inserted into an `external` method with the requisite name. This is already done in `ERC721.cairo`; however, additional functionality can be added. For instance, you could:
```

- Implement a pausing mechanism
- Add roles such as owner or minter

> - Modify the `transferFrom` function to perform actions [before](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol#L335)

Reviewer: ```suggestion
- Modify the `transferFrom` function to mimic the [`_beforeTokenTransfer` and `_afterTokenTransfer` hooks](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol#L335)
```

Reviewer: Maybe we can also point out an example, there's a few present in our presets such as pausing

Just be sure that the exposed `external` methods invoke their imported function logic a la `approve` invokes `ERC721_approve`. As an example, see below.

```python
from contracts.token.ERC721_base import ERC721_approve

@external
func approve{
        pedersen_ptr: HashBuiltin*, 
        syscall_ptr: felt*, 
        range_check_ptr
    }(to: felt, token_id: Uint256):
    ERC721_approve(to, token_id)
    return()
end

```

> 

Reviewer: I'd add a _Presets_ section to group the "ready to use" frontend contracts (i.e. what's below) with a brief description of the concept and maybe a mention under _Usage_.

### Mintable

The ERC721_Mintable preset offers a quick and easy setup for creating NFTs. This preset includes two imports: Ownable and ERC721_base. Further, this contract includes `mint` and `burn` methods to create and destroy NFTs. A common implementation of ERC721 contracts include setting a check on the `mint` and `burn` methods; where, only the owner can create and destroy tokens. In this preset, the owner is set in the constructor. 

### Metadata

> The Metadata extension is quite similar to the Mintable preset; however, this preset includes a baseURI parameter in the constructor and omits the `burn` method. Setting the baseURI initializes the `tokenURI` method to return an array with both the baseURI and the token_id. Without this set, `tokenURI` returns an array with a zero value. 

Reviewer: we should change this to mention we follow the Solidity implementation guideline of using the Metadata extension on all ERC721 implementations, since most people use it, but people can use the `_base` one directly which doesn't include it. (Although tbh I'm considering merging it to the base one too, to minimize boilerplate)

### Enumerable

> The Enumerable extension adds enumerability of all the token ids in the contract as well as all token ids owned by each account. This allows contracts to publish its full list of NFTs and make them discoverable. In regard to implementation, the extension exposes three new view methods as seen in the interface. Tokens are tracked via the `all_tokens_list` and `all_tokens_len`. No additional changes are required in the constructor for deployment.

Reviewer: Reading about a given preset it's hard with the explanation here and the interfaces above. Do you think it's worth putting them in here instead, or just linking from here?

## Utilities

### ERC721_Holder

Implementation of the `IERC721Receiver` interface.

Accepts all token transfers. Make sure the contract is able to use its token with `IERC721.safeTransferFrom`, `IERC721.approve` or `IERC721.setApprovalForAll`.

## API Specification

### IERC721 API

```jsx
func balanceOf(owner: felt) -> (balance: Uint256):
end

func ownerOf(token_id: Uint256) -> (owner: felt):
end

func safeTransferFrom(
        _from: felt, 
        to: felt, 
        token_id: Uint256, 
        data_len: felt,
        data: felt*
    ):
end

func transferFrom(_from: felt, to: felt, token_id: Uint256):
    end

func approve(approved: felt, token_id: Uint256):
end

func setApprovalForAll(operator: felt, approved: felt):
end

func getApproved(token_id: Uint256) -> (approved: felt):
end

func isApprovedForAll(owner: felt, operator: felt) -> (is_approved: felt):
end

```

#### `balanceOf`

Returns the number of tokens in `owner`'s account.

Parameters:

```jsx
owner: felt
```

Returns:

```jsx
balance: Uint256
```

#### `ownerOf`

> Returns the owner of the `token_id` token.

Reviewer: we should fix these to use camel case

Parameters:

```jsx
token_id: Uint256
```

Returns:

```jsx
owner: felt
```

#### `safeTransferFrom`

Safely transfers `token_id` token from `_from` to `to`, checking first that contract recipients are aware of the ERC721 protocol to prevent tokens from being forever locked.

> > Note this function should NOT be used. EOA differentiation still needs to be resolved on StarkNet.

Reviewer: should this be updated if we go after `is_account`?

Gives permission to `to` to transfer `token_id` token to another account. The approval is cleared when the token is transferred.

Parameters:

```jsx
to: felt
token_id: Uint256
```

Returns:

None.

#### `getApproved`

Returns the account approved for `token_id` token.

Parameters:

```jsx
token_id: Uint256
```

Returns:

```jsx
operator: felt
```

#### `setApprovalForAll`

Approve or remove `operator` as an operator for the caller. Operators can call `transferFrom` or `safeTransferFrom` for any token owned by the caller.

Parameters:

```jsx
operator: felt
```

Returns:

None.

#### `isApprovedForAll`

Returns if the `operator` is allowed to manage all of the assets of `owner`.

Parameters:

```jsx
owner: felt
operator: felt
```

Returns:

```jsx
is_approved: felt
```

---

### IERC721_Metadata API
```jsx
func name() -> (name: felt):
end

func symbol() -> (symbol: felt):
end

func tokenURI(token_id: Uint256) -> (uri_len: felt, uri: felt*):
end
```

#### `name`

Returns the token collection name.

Parameters:

None.

Returns:

```jsx
name: felt
```

#### `symbol`

Returns the token collection symbol.

Parameters: 

None.

Returns:

```jsx
symbol: felt
```

#### `tokenURI`

> Returns the Uniform Resource Identifier (URI) for `token_id` token. If the baseURI is not set, this function will return `[0]`. Otherwise, the returned array will return 3 values: baseURI, token_id.low, token_id.high. This differs from the Solidity implementation where the returned value is a concatenated string of both the baseURI and token ID.

Reviewer: needs to be updated

Parameters:

```jsx
index: Uint256
```

Returns:

```jsx
token_id: Uint256
```

#### `tokenOfOwnerByIndex`

Returns a token ID at a given `index` of all the tokens stored by the contract. Use along with [totalSupply](#totalSupply) to enumerate all tokens.

Parameters:

```jsx
owner: felt
index: Uint256
```

Returns:

```jsx
token_id: Uint256
```

---

### IERC721_Receiver API
```jsx

func onERC721Received(
        operator: felt, 
        _from: felt, 
        token_id: Uint256, 
        data_len: felt
        data: felt*
    ) -> (selector: felt):
end
```

#### `onERC721Received`

Interface for any contract that wants to support safeTransfers from ERC721 asset contracts.

Parameters:

```jsx
operator: felt
_from: felt
token_id: Uint256
data_len: felt
data: felt*
```

Returns:

```jsx
selector: felt
```
---

### IERC165 API
```jsx
func supportsInterface(interface_id: felt) -> (success: felt):
end
```

#### `supportsInterface`

Returns true if this contract implements the interface defined by `interface_id`.

Parameters:

```jsx
interface_id: felt
```

Returns:

```jsx

> success: felt

Reviewer: maybe this belongs to a 165-specific doc page? Or at least we should separate it enough so it's easy to move in a different PR.
